File: Data Structures & Algorithms/eating-bananas/submission-3.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)
        
        while left <= right:
            rate = (left + right) // 2
            time = 0

            for pile in piles:
                time += (pile + rate - 1) // rate
                if time > h:
                    break
            
            if time > h:
                left = rate + 1
            else:
                right = rate - 1
        
        return left

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    bool balanced = true;

    int maxDepth(TreeNode* root) {
        if (!root) {
            return 0;
        }

        int leftDepth = maxDepth(root->left);
        int rightDepth = maxDepth(root->right);
        if (abs(leftDepth - rightDepth) > 1) {
            balanced = false;
        }
        
        return 1 + max(leftDepth, rightDepth);
    }

public:
    bool isBalanced(TreeNode* root) {
        maxDepth(root);
        return balanced;
    }
};
